kubernetes/windows/main.ps1

```diff
@@ -289,35 +289,99 @@ function Start-Fluent {
+        Write-Host "Failed to set environment variable KUBERNETES_SERVICE_PORT for target 'machine' since it is either null or empty"
+    }
+    
+    $nodeIp = [System.Environment]::GetEnvironmentVariable("NODE_IP", "process")
+    if (![string]::IsNullOrEmpty($nodeIp)) {
+        [System.Environment]::SetEnvironmentVariable("NODE_IP", $nodeIp, "machine")
+        Write-Host "Successfully set environment variable NODE_IP - $($nodeIp) for target 'machine'..."
+    }
+    else {
+        Write-Host "Failed to set environment variable NODE_IP for target 'machine' since it is either null or empty"
+    }
 
-function Bootstrap-CACertificates {
+    Write-Host "Installing telegraf service"
+    C:\opt\telegraf\telegraf.exe --service install --config "C:\etc\telegraf\telegraf.conf"
+
+    # Setting delay auto start for telegraf since there have been known issues with windows server and telegraf -
+    # https://github.com/influxdata/telegraf/issues/4081
+    # https://github.com/influxdata/telegraf/issues/3601
     try {
-        # This is required when the root CA certs are different for some clouds.
-        $caCerts=Invoke-WebRequest 'http://168.63.129.16/machine?comp=acmspackage&type=cacertificates&ext=json' -UseBasicParsing | ConvertFrom-Json
-        if (![string]::IsNullOrEmpty($caCerts)) {
-            $certificates = $caCerts.Certificates
-            for ($index = 0; $index -lt $certificates.Length ; $index++) {
-                $name=$certificates[$index].Name
-                $certificates[$index].CertBody > $name
-                Write-Host "name: $($name)"
-                Import-Certificate -FilePath .\$name  -CertStoreLocation 'Cert:\LocalMachine\Root' -Verbose
-            }
+        $serverName = [System.Environment]::GetEnvironmentVariable("PODNAME", "process")
+        if (![string]::IsNullOrEmpty($serverName)) {
+            sc.exe \\$serverName config telegraf start= delayed-auto
+            Write-Host "Successfully set delayed start for telegraf"
+
+        } else {
+            Write-Host "Failed to get environment variable PODNAME to set delayed telegraf start"
         }
     }
     catch {
-        $e = $_.Exception
-        Write-Host $e
-        Write-Host "exception occured in Bootstrap-CACertificates..."
+            $e = $_.Exception
+            Write-Host $e
+            Write-Host "exception occured in delayed telegraf start.. continuing without exiting"
     }
+    Write-Host "Running telegraf service in test mode"
+    C:\opt\telegraf\telegraf.exe --config "C:\etc\telegraf\telegraf.conf" --test
+    Write-Host "Starting telegraf service"
+    C:\opt\telegraf\telegraf.exe --service start
+
+    # Trying to start telegraf again if it did not start due to fluent bit not being ready at startup
+    Get-Service telegraf | findstr Running
+    if ($? -eq $false)
+    {
+        Write-Host "trying to start telegraf in again in 30 seconds, since fluentbit might not have been ready..."
+        Start-Sleep -s 30
```

Comment:
rather than sleep, probable checking for fluentbit ready or not is right thing to do.

Reply:
> rather than sleep, probable checking for fluentbit ready or not is right thing to do.
> 
> Rashmi - We dont want to wait indefinitely if fluentbit doesnt start. We retry once and then move on. We can discuss if there are concerns.

I am not saying trying indefinitely. rather poll for fluent bit start with 1 sec delay and if its gets start quickly, telegraf will start in worst case, you will be polling 30 sec or whatever threshold interval. This way, majority of the cases no impact of 30 sec container start/initial ingestion delay.